axum-extra/src/middleware/middleware_fn.rs

```diff
@@ -0,0 +1,261 @@
+//! Create middleware from async functions.
+//!
+//! See [`from_fn`] for more details.
+
+use axum::{
+    body::{self, Bytes, HttpBody},
+    response::{IntoResponse, Response},
+    BoxError,
+};
+use http::Request;
+use pin_project_lite::pin_project;
+use std::{
+    any::type_name,
+    convert::Infallible,
+    fmt,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{util::BoxCloneService, ServiceBuilder};
+use tower_http::ServiceBuilderExt;
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Create a middleware from an async function.
+///
+/// `from_fn` requires the function given to
+///
+/// 1. Be an `async fn`.
+/// 2. Take [`Request`](http::Request) as the first argument.
+/// 3. Take [`Next<B>`](Next) as the second argument.
+/// 4. Return something that implements [`IntoResponse`].
+///
+/// # Example
+///
+/// ```rust
+/// use axum::{
+///     Router,
+///     http::{Request, StatusCode},
+///     routing::get,
+///     response::IntoResponse,
+/// };
+/// use axum_extra::middleware::{self, Next};
+///
+/// async fn auth<B>(req: Request<B>, next: Next<B>) -> impl IntoResponse {
+///     let auth_header = req.headers().get(http::header::AUTHORIZATION);
+///
+///     match auth_header {
+///         Some(auth_header) if auth_header == "secret" => {
+///             Ok(next.run(req).await)
+///         }
+///         _ => Err(StatusCode::UNAUTHORIZED),
+///     }
+/// }
+///
+/// let app = Router::new()
+///     .route("/", get(|| async { /* ... */ }))
+///     .route_layer(middleware::from_fn(auth));
+/// # let app: Router = app;
+/// ```
+pub fn from_fn<F>(f: F) -> MiddlewareFnLayer<F> {
+    MiddlewareFnLayer { f }
+}
+
+/// A [`tower::Layer`] from an async function.
+///
+/// [`tower::Layer`] is used to apply middleware to [`axum::Router`]s.
+///
+/// Created with [`from_fn`]. See that function for more details.
+#[derive(Clone, Copy)]
+pub struct MiddlewareFnLayer<F> {
+    f: F,
+}
+
+impl<S, F> Layer<S> for MiddlewareFnLayer<F>
+where
+    F: Clone,
+{
+    type Service = MiddlewareFn<F, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        MiddlewareFn {
+            f: self.f.clone(),
+            inner,
+        }
+    }
+}
+
+impl<F> fmt::Debug for MiddlewareFnLayer<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MiddlewareFnLayer")
+            .field("f", &format_args!("{}", type_name::<F>()))
```

Comment:
I didn't know that `fmt::Arguments` implements `Debug` in addition to `Display`. Maybe add a comment though? I still find it looks pretty weird.

```suggestion
            // Write out the type name, without quoting it as `&type_name::<F>()` would
            .field("f", &format_args!("{}", type_name::<F>()))
```